src/main/scala/com/danielasfregola/twitter4s/TwitterRestClient.scala

```diff
@@ -24,19 +25,19 @@ import com.danielasfregola.twitter4s.http.clients.rest.suggestions.TwitterSugges
 import com.danielasfregola.twitter4s.http.clients.rest.trends.TwitterTrendClient
 import com.danielasfregola.twitter4s.http.clients.rest.users.TwitterUserClient
 import com.danielasfregola.twitter4s.util.Configurations._
-import com.danielasfregola.twitter4s.util.SystemShutdown
+import com.danielasfregola.twitter4s.util.{Configurations, SystemShutdown}
 
 /** Represents the functionalities offered by the Twitter REST API
   */
-class TwitterRestClient(val consumerToken: ConsumerToken, val accessToken: AccessToken)(implicit _system: ActorSystem =
-                                                                                          ActorSystem("twitter4s-rest"))
+class TwitterRestClient(val consumerToken: ConsumerToken,
```

Comment:
I think this constructor is the main problem in this PR - I believe by addressing this most of the ugliness will go away.

I would suggest passing the client as its parameter and use the tokens only in the `apply` methods.

This would look like something like this (disclaimer...I am just writing pseudo code, I do not guarantee it will compile):

```
class TwitterRestClient private (restClient: RestClient)(implicit _system: ActorSystem = ActorySystem("...")) { ... } // private ensures that only apply methods can call its constructor

object TwitterRestClient {
     
    def apply(consumerToken, accessToken) = {
        val client = // get the client for those tokens
        new TwitterRestClient(client)
   }

     def apply(bearerToken) = {
        val client = // get the bearer client
        new TwitterRestClient(client)
   }

   def apply() = {
       // check the env variables:
       // if only consumer and access tokens, delegate to their apply method
      // if only bearer token, delegate to the apply
     // if all three, delegate to another apply method....which you can create usin the same pattern
     // else....throw an exception
   }
```

Also a tip on using options:
```
val consumerToken: Option[String] = ???
val accessToken: Option[String] = ???
val bearerToken: Option[String] = ???
(consumerToken, accessToken, bearerToken) match {
    case (Some(consumer), Some(access), Some(bearer)) => // all three are present!
    case (Some(consumer), Some(access), None) => // only consumer and access tokens are there
    case  (None, None, Some(bearer)) => // only bearer is there
    case _ => // none of the above, we should error
}
```


....I hope this helps! :)

}

Reply:
This does help! I was going to say "your explanations are really clear and concise, you could write a book on Scala" and then I realized you already did :smile: 

I think I see how I can make my changes a little nicer, I'll try out these suggestions and see where they get me. A lot of this is me jumping between 4 languages thanks to school and work, so language-specific feature's like Scala's `Option` usually does not come to mind as I develop. Thanks again for the detailed feedback.